File: src/requests.lua
---Global table for HTTP requests.
requests = {}

---Create a new requests client for POST, PUT, PATCH, DELETE operations.
---@return table The requests client.
function requests.create_client()
    return {
        post = function(url, body)
        end,
        put = function(url, body)
        end,
        patch = function(url, body)
        end,
        delete = function(url)
        end
    }
end

---Perform a GET request to a URL.
---@param url string The URL to request.
---@return table The response object.
function requests.get(url)
end

---Get the status of the response.
---@param response table The response object.
---@return number The HTTP status code.
function requests.get_status(response)
end

---Get the headers of the response.
---@param response table The response object.
---@return table A table containing headers.
function requests.get_headers(response)
end

---Get a specific header value from the response.
---@param response table The response object.
---@param key string The header key to retrieve.
---@return string|nil The value of the header, or nil if not found.
function requests.get_header_value(response, key)
end

---Get the response body as text.
---@param response table The response object.
---@return string The response body as a string.
function requests.get_text(response)
end

---Get the response body as JSON.
---@param response table The response object.
---@return table The response body parsed as JSON.
function requests.get_json(response)
end

File: src/requests.rs
use log::{debug, info};
use mlua::{Lua, Result};

async fn get(_lua: &Lua, url: String) -> Result<String> {
    info!("GET {}", url);
    let resp = reqwest::get(&url).await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

async fn post(_lua: &Lua, url: String, body: String) -> Result<String> {
    info!("POST {}", url);
    let client = reqwest::Client::new();
    let resp = client.post(&url).body(body).send().await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

async fn put(_lua: &Lua, url: String, body: String) -> Result<String> {
    info!("PUT {}", url);
    let client = reqwest::Client::new();
    let resp = client.put(&url).body(body).send().await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

async fn delete(_lua: &Lua, url: String) -> Result<String> {
    info!("DELETE {}", url);
    let client = reqwest::Client::new();
    let resp = client.delete(&url).send().await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

async fn patch(_lua: &Lua, url: String, body: String) -> Result<String> {
    info!("PATCH {}", url);
    let client = reqwest::Client::new();
    let resp = client.patch(&url).body(body).send().await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

async fn head(_lua: &Lua, url: String) -> Result<String> {
    info!("HEAD {}", url);
    let client = reqwest::Client::new();
    let resp = client.head(&url).send().await.unwrap();

    info!("Status: {}", resp.status());
    let body = resp.text().await.unwrap();

    debug!("<< {:?}", body);
    Ok(body)
}

pub fn inject_module(lua: &Lua) -> Result<()> {
    let m = lua.create_table()?;
    m.set(
        "get",
        lua.create_async_function(|_lua: Lua, url: String| async move { get(&_lua, url).await })?,
    )?;

    m.set(
        "post",
        lua.create_async_function(|_lua: Lua, (url, body): (String, String)| async move {
            post(&_lua, url, body).await
        })?,
    )?;

    m.set(
        "put",
        lua.create_async_function(|_lua: Lua, (url, body): (String, String)| async move {
            put(&_lua, url, body).await
        })?,
    )?;

    m.set(
        "delete",
        lua.create_async_function(
            |_lua: Lua, url: String| async move { delete(&_lua, url).await },
        )?,
    )?;

    m.set(
        "patch",
        lua.create_async_function(|_lua: Lua, (url, body): (String, String)| async move {
            patch(&_lua, url, body).await
        })?,
    )?;

    m.set(
        "head",
        lua.create_async_function(|_lua: Lua, url: String| async move { head(&_lua, url).await })?,
    )?;

    lua.globals().set("requests", m)?;
    Ok(())
}
